onnxruntime/core/providers/webnn/builders/impl/expand_op_builder.cc

```diff
@@ -0,0 +1,126 @@
+  if (!status.IsOK()) {
+    LOGS(logger, ERROR) << "Error while unpacking shape: " << status.ErrorMessage();
+    return false;
+  }
+  const auto& dims = tensor.dims();
+  if (dims.empty() || dims[0] == 0) {
+    LOGS(logger, VERBOSE) << "The shape of expand cannot be empty.";
+    return false;
+  }
+  if (dims.size() != 1) {
+    LOGS(logger, VERBOSE) << "The shape of expand must be 1D.";
+    return false;
+  }
+  const int64_t* shape_data = reinterpret_cast<const int64_t*>(unpacked_tensor.data());
+  shape = std::vector<int64_t>{shape_data, shape_data + dims[0]};
+  return true;
+}
+
+// Add operator related.
+
+void ExpandOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  model_builder.AddInitializerToSkip(node.InputDefs()[1]->Name());
+}
+
+Status ExpandOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                              const Node& node,
+                                              const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& initializers(model_builder.GetInitializerTensors());
+  const auto& shape_tensor = *initializers.at(input_defs[1]->Name());
+  std::vector<int64_t> raw_shape;
+  // We already checked the raw_shape (new_shape) in IsOpSupportedImpl.
+  GetExpandShape(shape_tensor, raw_shape, logger);
+  std::vector<int32_t> new_shape;
+  std::transform(raw_shape.cbegin(), raw_shape.cend(),
+                 std::back_inserter(new_shape),
+                 [](int64_t dim) -> int32_t { return SafeInt<int32_t>(dim); });
+  emscripten::val input = model_builder.GetOperand(input_defs[0]->Name());
+  emscripten::val output =
+      model_builder.GetBuilder().call<emscripten::val>("expand",
+                                                       input, emscripten::val::array(new_shape));
+  model_builder.AddOperand(node.OutputDefs()[0]->Name(), std::move(output));
+  return Status::OK();
+}
+
+// Operator support related.
+
+bool ExpandOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                                        const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& shape_name = input_defs[1]->Name();
+  if (!Contains(initializers, shape_name)) {
+    LOGS(logger, VERBOSE) << "The shape must be a constant initializer";
+    return false;
+  }
+
+  std::vector<int64_t> new_shape;
+  const auto& shape_tensor = *initializers.at(shape_name);
+  if (!GetExpandShape(shape_tensor, new_shape, logger)) {
+    LOGS(logger, VERBOSE) << "Cannot get shape.";
+    return false;
+  }
+
+  std::vector<int64_t> input_shape;
+  if (!GetShape(*input_defs[0], input_shape, logger)) {
+    LOGS(logger, VERBOSE) << "Cannot get input's shape.";
+    return false;
+  }
+
+  if (input_shape.empty()) {
+    LOGS(logger, VERBOSE) << "Expand does not support empty input's shape";
```

Comment:
Technically this is not illegal. It just follows broadcasting rules for scalars. You can try it with the ORT CPU EP and see it succeeds:

```json
    {
      "op_type": "Expand",
      "input": [0, 1, 2, 3, 4],
      "shape": [],
      "output": [0, 1, 2, 3, 4],
      "T": "float32"
    },
```

![image](https://github.com/Honry/onnxruntime/assets/1809166/1dc0a0c1-2a1d-4436-a87b-71ee67db55f9)